### PR TITLE
ZQS 671 Dynamic circuit estimation task factory

### PR DESCRIPTION
**Pull request by laurgao**
3rd PR in cost function refactor

Merge after #406

### CODE REVIEW

File: src/python/zquantum/core/cost_function.py [hunk 1]
```diff
@@ -429,3 +429,54 @@ def _tasks_factory(parameters: np.ndarray) -> List[EstimationTask]:
...
+        An EstimationTasksFactory object.
+    """
+
+    def _tasks_factory(parameters: np.ndarray) -> List[EstimationTask]:
+
+        # NOTE: `ansatz._generate_circuit(parameters)` currently does not produce an
+        # executable circuit, but rather a parametrized circuit with sympy symbols in
+        # some ansatz implementations. (Ex. see ansatzes in z-quantum-qaoa)
+        #
+        # Combined with how this is a private method, we will probably have to somewhat
```
Comment (mstechly) [inline src/python/zquantum/core/cost_function.py:464]: I would write TODO here and perhaps add a story to Jira :)

File: src/python/zquantum/core/interfaces/ansatz.py [hunk 2]
```diff
@@ -87,6 +87,9 @@ def _generate_circuit(self, params: Optional[np.ndarray] = None) -> Circuit:
         Will return parametrized circuits if no parameters are passed and the ansatz
         supports parametrized circuits.
 
+        Ansatzes that implement this method must not ignore parameters if parameters
```
Comment (mstechly) [inline src/python/zquantum/core/interfaces/ansatz.py:90]: I think this is a pretty logical thing, so I'm not really sure if it's worth mentioning here.

**Review by mstechly — APPROVED** [review]

File: src/python/zquantum/core/interfaces/ansatz.py (hunk 2 above)
Comment (laurgao) [reply src/python/zquantum/core/interfaces/ansatz.py:90]: Some ansatzes don't do it though 🤔

Comment (mstechly) [reply src/python/zquantum/core/interfaces/ansatz.py:90]: Then leave it be 🤷

**Pull request merged**